### PR TITLE
FOGL-8306: Alerts stuff (patch2)

### PR DESCRIPTION
**Pull request by MonikaSharma06**
- [x] Group by urgency, group itself sorted by ts (list by urgency group)
- [x] Delete last remaining alert, an empty dropdown remains until the next ping; and bell icon disappears. Bell icon should not disappear but badge should be removed and that empty dropdown. Bell icon hover title in this case should be No Alerts
- [x] Feedback from previous PR
- [x] If message is longer then 3 lines then ellipsis ...
- [x] At bottom after divider if alerts length > 1, Add Hide all ghost button in right and Sort by Timestamp or Sort by Urgency (toggle) in the left

### CODE REVIEW

File: src/app/components/core/system-alert/system-alert.component.ts [hunk 1]
```diff
@@ -58,6 +87,24 @@ export class SystemAlertComponent {
...
+      this.sortByKey = 'time';
+    }
+  }
+
+  sortAlertsByTimestamp(alerts) {
+    return alerts.sort((a, b) => new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime());
+  }
+
+  sortAlertsByUrgency() {
+    return this.systemAlerts.sort((a, b) => a.urgency.localeCompare(b.urgency));
```
Comment (praveen-garg) [inline src/app/components/core/system-alert/system-alert.component.ts:105]: No

    const urgencyOrder = { "Critical": 0, "High": 1, "Normal": 2, "Low": 3 };
    
    // Group by urgency
    const groupedByUrgency = data.reduce((acc, item) => {
        if (!res[item.urgency]) {
            res[item.urgency] = [];
        }
        res[item.urgency].push(item);
        return res;
    }, {});

    // Sort urgency keys based on urgencyOrder
    const sortedUrgencyKeys = Object.keys(groupedByUrgency).sort((a, b) => urgencyOrder[a] - urgencyOrder[b]);
    
    // Sort each group by timestamp
    sortedUrgencyKeys.forEach(key => {
        groupedByUrgency[key].sort((a, b) => new Date(a.timestamp) - new Date(b.timestamp));
    });

File: src/app/components/core/system-alert/system-alert.component.ts [hunk 2]
```diff
@@ -97,7 +144,30 @@ export class SystemAlertComponent {
...
+  }
+
+  deleteAllAlerts() {
+    this.ngProgress.start();
+    this.systemAlertService.deleteAllAlerts().
+    subscribe(
+      (data: any) => {
+        this.ngProgress.done();
+        this.reenableButton.emit(false);
+        this.alertsCount = 0;
```
Comment (praveen-garg) [inline src/app/components/core/system-alert/system-alert.component.ts:170]: also set this.systemAlerts = []?

**Review by praveen-garg — CHANGES_REQUESTED** [review]

File: src/app/components/core/system-alert/system-alert.component.ts [hunk 3]
```diff
@@ -58,6 +71,44 @@ export class SystemAlertComponent {
     );
   }
 
+  sortByTimestamp() {
+    this.sortByKey = 'urgency';
```
Comment (praveen-garg) [inline src/app/components/core/system-alert/system-alert.component.ts:75]: Add an inline comment that its toggling for next action text

File: src/app/components/core/system-alert/system-alert.component.ts [hunk 4]
```diff
@@ -58,6 +71,44 @@ export class SystemAlertComponent {
...
+    return this.systemAlerts.sort((a, b) => new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime());
+  }
+
+  groupByUrgencySortedByTime(alerts = []) {
+    if (alerts.length === 0) {
```
Comment (praveen-garg) [inline src/app/components/core/system-alert/system-alert.component.ts:80]: Add inline comment and explain why is this required?

File: src/app/components/core/system-alert/system-alert.component.ts [hunk 5]
```diff
@@ -58,6 +71,44 @@ export class SystemAlertComponent {
...
+    [...groupedByUrgencyObj[key]]
+    );
+
+    // Sort each group by timestamp
+    groupedByUrgency.forEach(alerts => {
+      alerts.sort((a, b) => new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime());
+    });
+    
+    const systemAlerts = groupedByUrgency.reduce((a, value) => a.concat(value), []);
+    this.sortByKey = 'time';
```
Comment (praveen-garg) [inline src/app/components/core/system-alert/system-alert.component.ts:107]: Add toggling next possible action text

**Review by praveen-garg — DISMISSED** [review]

**Review by mshariq-nerd — APPROVED** [review]

File: src/app/components/core/system-alert/system-alert.component.ts (hunk 4 above)
Comment (praveen-garg) [reply src/app/components/core/system-alert/system-alert.component.ts:84]: I'm still not clear on this.

File: src/app/components/core/system-alert/system-alert.component.ts [hunk 6]
```diff
@@ -58,6 +72,49 @@ export class SystemAlertComponent {
...
+    .map((key) => 
+    [...groupedByUrgencyObj[key]]
+    );
+
+    // Sort each group by timestamp
+    groupedByUrgency.forEach(alerts => {
+      alerts.sort((a, b) => new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime());
+    });
+    
+    const systemAlerts = groupedByUrgency.reduce((a, value) => a.concat(value), []);
```
Comment (praveen-garg) [inline src/app/components/core/system-alert/system-alert.component.ts:110]: What is the use of this? and can be directly assigned to  this.systemAlerts = ... if required? Add an inline comment

File: src/app/components/core/system-alert/system-alert.component.ts [hunk 7]
```diff
@@ -58,6 +72,49 @@ export class SystemAlertComponent {
...
+    // toggle the next possible button text
+    this.sortByKey = 'time';
+    
+    this.systemAlerts = systemAlerts;
+    this.hideLoadingSpinner();
```
Comment (praveen-garg) [inline src/app/components/core/system-alert/system-alert.component.ts:115]: hideLoadingSpinner() should not be here.

**Pull request merged**